### PR TITLE
sound: check DEVICE_FORM_FACTOR property

### PR DESCRIPTION
**Pull request by kevinmost**
DEVICE_FORM_FACTOR, if present, will be used to determine whether the
headphones icon is used. If it is not present, the existing active_port
method is used to determine if the headphones icon is used.

### CODE REVIEW

**Comment by codicodi** [comment]
Looks alright. FWIW [Waybar](https://github.com/Alexays/Waybar/blob/ad5ea7ad2b698c7a60c6bb7e81f06a662cdba819/src/modules/pulseaudio.cpp#L214-L237) folks seem to also check both form factor and port name

**Comment by kevinmost** [comment]
Great, thanks for looking

File: src/blocks/sound.rs [hunk 1]
```diff
@@ -653,6 +663,10 @@ impl SoundDevice for PulseAudioSoundDevice {
         self.active_port.clone()
     }
 
+    fn form_factor(&self) -> Option<String> {
+        self.form_factor.clone()
```
Comment (MaxVerevkin) [inline src/blocks/sound.rs:667]: This function can return `&str` in order to avoid `clone()`

File: src/blocks/sound.rs [hunk 2]
```diff
@@ -653,6 +663,10 @@ impl SoundDevice for PulseAudioSoundDevice {
         self.active_port.clone()
```
Comment (MaxVerevkin) [inline src/blocks/sound.rs:663]: Not part of your PR, but can you please remove this `clone()` too?

File: src/blocks/sound.rs (hunk 1 above)
Comment (kevinmost) [reply src/blocks/sound.rs]: How's this?

File: src/blocks/sound.rs (hunk 2 above)
Comment (kevinmost) [reply src/blocks/sound.rs]: Ditto

File: src/blocks/sound.rs [hunk 3]
```diff
@@ -825,9 +840,27 @@ impl Default for SoundConfig {
 }
 
 impl Sound {
-    fn icon(&self, volume: u32, headphones: bool) -> String {
-        if self.headphones_indicator && self.device_kind == DeviceKind::Sink && headphones {
-            return String::from("headphones");
+    fn icon(&self, volume: u32) -> String {
+        if self.headphones_indicator && self.device_kind == DeviceKind::Sink {
+            let headphones = match self.device.form_factor().as_deref() {
```
Comment (MaxVerevkin) [inline src/blocks/sound.rs:845]: Now you can remove this `as_deref()` as it doesn't do anything.

File: src/blocks/sound.rs [hunk 4]
```diff
@@ -825,9 +840,27 @@ impl Default for SoundConfig {
...
+                Some("headset") | Some("headphone") | Some("hands-free") | Some("portable") => true,
+                // Per discussion at
+                // https://github.com/greshake/i3status-rust/pull/1363#issuecomment-1046095869,
+                // some sinks may not have the form_factor property, so we should fall back to the
+                // active_port if that property is not present.
+                None => self
+                    .device
+                    .active_port()
+                    .map(|p| p.contains("headphones"))
+                    .unwrap_or(false),
```
Comment (MaxVerevkin) [inline src/blocks/sound.rs:857]: Here you can use `.map_or(false, |p| p.contains("headphones"))`

**Comment by kevinmost** [comment]
Done, thanks.

**Review by MaxVerevkin — APPROVED** [review]

**Comment by MaxVerevkin** [comment]
Thanks!

**Pull request merged**